pkg/controller/issueapi/logic.go

```diff
@@ -0,0 +1,305 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package issueapi
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/google/exposure-notifications-server/pkg/logging"
+	"github.com/google/exposure-notifications-server/pkg/timeutils"
+	"github.com/google/exposure-notifications-verification-server/internal/project"
+	"github.com/google/exposure-notifications-verification-server/pkg/api"
+	"github.com/google/exposure-notifications-verification-server/pkg/controller"
+	"github.com/google/exposure-notifications-verification-server/pkg/database"
+	"github.com/google/exposure-notifications-verification-server/pkg/observability"
+	"github.com/google/exposure-notifications-verification-server/pkg/otp"
+	"github.com/google/exposure-notifications-verification-server/pkg/sms"
+	"go.opencensus.io/stats"
+)
+
+func (c *Controller) issue(ctx context.Context, request *api.IssueCodeRequest,
```

Comment:
This function signature is a bit weird because it's unclear what's required and what's option. For example, `db` and `realm` are required - passing `nil` will panic. But `user` and `authApp` are optional and can both be nil.

At the very least, we should document this, but it would be good to add some runtime checks or potentially change the signature to make this more obvious.

Reply:
Changed the signature - let's pull what we can off of context instead of passing them all around. Esp if we can assume they already checked/rejected by middleware